[GSoC'24] M1.5: Add new classrooms page

AFZL210 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #19849 .
2. This PR does the following: Added a new classrooms page and updated the routing.

**Routing flow:** If we have just one classroom or if the multiple classrooms feature is disabled then we redirect the learner to the existing math classroom.
- If we have no public classroom and the environment is DEV then we show the classrooms page with banner saying "To create classrooms, visit classroom admin page"
- and if its PROD then we show 404 page
![image](https://github.com/oppia/oppia/assets/79896602/e733a278-a7b3-4e56-a2be-8b7aee075153)

- On the classrooms page we show all the public classrooms card and we also show a classroom under construction card if we have **atleast one** private classroom.

[Figma link](https://www.figma.com/design/CxEKn4FvaZRAAbvuqDchD1/Oppia-%23109-Oppia-Multiple-Classrooms?node-id=1309-21355&t=o1XcJJUKpXKGkLFu-4)

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [ ] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

https://github.com/oppia/oppia/assets/79896602/1d14233b-f52f-41c4-9df5-c748f29dbd26



If we have 0 classrooms in DEV_MODE We show a banner which tells user from where they can create classrooms.
![Screenshot from 2024-06-22 16-05-06](https://github.com/oppia/oppia/assets/79896602/538e33e9-eb40-4ddc-9219-9ee242a6c430)


<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network
N/A
<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone
![image](https://github.com/oppia/oppia/assets/79896602/dc5ae7fe-8e2c-46ad-ad82-5e5633930b97)

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language
![Screenshot from 2024-06-22 02-24-57](https://github.com/oppia/oppia/assets/79896602/ab94d4d6-43bc-4a1f-bfb3-4bc744c1fc9e)

As the PR is getting quite big and we will be implementing functions to get translation keys for classroom data in M1.7 anyways so we will update translation for M1.5 one M1.7 is complete. I have filed an issue to track this https://github.com/oppia/oppia/issues/20518

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

AFZL210 committed:
Fix part of #19849: Add new classrooms page

AFZL210 committed:
Merge branch 'develop' into fix#19849-classrooms-page

AFZL210 committed:
add mobile breadcrumbs

AFZL210 committed:
add i18n for classrooms description

AFZL210 commented:
@nikitaevg @vojtechjelinek PTAL.

S4v8n commented:
@AFZL210 It is coming along nicely, thank you for your work on this page.   There are a few minor tweaks: 

For the web version: 

1. On the main nav bar, the heading 'Home' should not be there.  Please remove it. 
2. On the main nav bar, the headings should be spaced 24px apart and the heading 'Donate' should be spaced 32px from the language selector CTA button. Please double check. 
3. The breadcrumbs alignment needs to align with the title heading below it. Currently it is a little further to the left of the title heading. 
4. "Oppia Classrooms" title needs to be 28px from the graphic image below. Currently it looks close to each other. 
5. "Step into...." subtext needs to be left aligned with the Title heading and the classroom subject cards.  Please double check the font size 18, and medium. 
6. Subject cards titles, first letter needs to be capitalized.  
7. The math card subtext size need to be 158px, the word 'fun' should be on the second line. 
8. The Science card needs the remaining subtext. 
9. Coming soon card, title needs to align with the card titles. 
10. on all three subject cards, the spacing between icon to title is 32px, title to subtext is 24px.  Please adjust accordingly. 
11. the spacing between the subject cards and the search bar should 52px. 
12. The search bar is not the updated version, will the new version be implemented on this round? 

For iPad version: 

13. The coming soon card looks smaller than the rest of the subject cards

For the iPhone SE version: 

14. the top nav bar doesn't look aligned, there is white space showing on the left side.
15. the subject cards need to be 318px wide by 354px height.  


Those are all the items I noticed on this round.  This should be to the mobile version as well. 
Thank you again for all your hard work on this!   - Best, Kurin

AFZL210 commented:
> @AFZL210 It is coming along nicely, thank you for your work on this page. There are a few minor tweaks:
> 
> For the web version:
> 
> 1. On the main nav bar, the heading 'Home' should not be there.  Please remove it.
> 2. On the main nav bar, the headings should be spaced 24px apart and the heading 'Donate' should be spaced 32px from the language selector CTA button. Please double check.
> 3. The breadcrumbs alignment needs to align with the title heading below it. Currently it is a little further to the left of the title heading.
> 4. "Oppia Classrooms" title needs to be 28px from the graphic image below. Currently it looks close to each other.
> 5. "Step into...." subtext needs to be left aligned with the Title heading and the classroom subject cards.  Please double check the font size 18, and medium.
> 6. Subject cards titles, first letter needs to be capitalized.
> 7. The math card subtext size need to be 158px, the word 'fun' should be on the second line.
> 8. The Science card needs the remaining subtext.
> 9. Coming soon card, title needs to align with the card titles.
> 10. on all three subject cards, the spacing between icon to title is 32px, title to subtext is 24px.  Please adjust accordingly.
> 11. the spacing between the subject cards and the search bar should 52px.
> 12. The search bar is not the updated version, will the new version be implemented on this round?
> 
> For iPad version:
> 
> 13. The coming soon card looks smaller than the rest of the subject cards
> 
> For the iPhone SE version:
> 
> 14. the top nav bar doesn't look aligned, there is white space showing on the left side.
> 15. the subject cards need to be 318px wide by 354px height.
> 
> Those are all the items I noticed on this round. This should be to the mobile version as well. Thank you again for all your hard work on this! - Best, Kurin

Thank you Kurin, I will fix those issues and assign you again once it's fixed.

AFZL210 committed:
fix style

AFZL210 committed:
Merge branch 'develop' into fix#19849-classrooms-page

AFZL210 committed:
Merge branch 'develop' into fix#19849-classrooms-page

AFZL210 committed:
fix style

AFZL210 commented:
@S4v8n 1, 2 - It's part of the main navbar, so it's kind of unrelated to my change.
12 - We need to update the search bar too. I will file an issue for it.

I have fixed all the major issues, and this page will be live on 28 June on a test server. Then I guess we can find more issues and fix them.

I have updated the video PTAL. Thanks

AFZL210 commented:
@nikitaevg @vojtechjelinek PTAL.

S4v8n commented:
> @S4v8n 1, 2 - It's part of the main navbar, so it's kind of unrelated to my change. 12 - We need to update the search bar too. I will file an issue for it.
> 
> I have fixed all the major issues, and this page will be live on 28 June on a test server. Then I guess we can find more issues and fix them.
> 
> I have updated the video PTAL. Thanks

@AFZL210 its looking really good.  I only noticed a couple of small alignment issues

For the desktop: 

The three subject cards, although they are aligned when you did the code checks, visually the left card is not aligned, it looks like it is offset by a few px. 


For the mobile: 

the subject cards are not centered, and should align with the left side of the subtext.  

Those are the comments I have thus far, once the page is on the test server I can review it in greater detail.  

Thank you again, have a great weekend!  It's looking really good.

AFZL210 committed:
update routing

AFZL210 commented:
> > @S4v8n 1, 2 - It's part of the main navbar, so it's kind of unrelated to my change. 12 - We need to update the search bar too. I will file an issue for it.
> > I have fixed all the major issues, and this page will be live on 28 June on a test server. Then I guess we can find more issues and fix them.
> > I have updated the video PTAL. Thanks
> 
> @AFZL210 its looking really good. I only noticed a couple of small alignment issues
> 
> For the desktop:
> 
> The three subject cards, although they are aligned when you did the code checks, visually the left card is not aligned, it looks like it is offset by a few px.
> 
> For the mobile:
> 
> the subject cards are not centered, and should align with the left side of the subtext.
> 
> Those are the comments I have thus far, once the page is on the test server I can review it in greater detail.
> 
> Thank you again, have a great weekend!

Thank you Kurin. Yes it does look a little off but I guess it's due to the shadow effect because when I check the alignemt it looks fine. we can fix it once it's live on some test server.

nikitaevg commented on core/controllers/access_validators.py:
```diff
@@ -80,6 +80,40 @@ def get(self) -> None:
+        """
+        classrooms = classroom_config_services.get_all_classrooms()
+        have_atleast_one_public_classroom = False
+        for classroom in classrooms:
+            if classroom.is_published:
+                have_atleast_one_public_classroom = True
+                break
```
I think you can write it using Python list processing functions. Something like any(classrooms, lambda x: x.is_published).

Also, the variable name is a bit cumbersome. have->has, and atleast_one is implied, so I'd say it's has_public_classrooms

AFZL210 replied:
Done.

nikitaevg commented on core/controllers/access_validators.py:
```diff
@@ -80,6 +80,40 @@ def get(self) -> None:
+                have_atleast_one_public_classroom = True
+                break
+
+        if not feature_flag_services.is_feature_flag_enabled(
+            feature_flag_list.FeatureNames.ENABLE_MULTIPLE_CLASSROOMS.value,
+            user_id=self.user_id
+        ) or (not have_atleast_one_public_classroom and not constants.DEV_MODE):
```
Too many not-s, I find it difficult to understand what's going on here.

Also, feature flag check might be moved to the top of the function. Then it's `if not has_public_classrooms and not constants.DEV_MODE`, which is fine

AFZL210 replied:
Done.

nikitaevg commented on core/templates/pages/classrooms-page/classrooms-page.component.ts:
```diff
@@ -0,0 +1,86 @@
+        this.classroomSummaries = data;
+        for (let i = 0; i < this.classroomSummaries.length; i++) {
+          if (this.classroomSummaries[i].is_published) {
+            this.publicClassroomCount += 1;
+            this.singlePublicClassroomUrlFragment =
```
These fields can be variables, not fields of the class, right? It's better to keep a class state as small as possible, less things to worry about when working with that class

AFZL210 replied:
yes makes sense. removed them.

nikitaevg commented on core/templates/pages/classrooms-page/classrooms-page-root.component.html:
```diff
@@ -0,0 +1,9 @@
+<oppia-angular-root>
```
Why do you need this component? What if you merge it with the classrooms-page?

AFZL210 replied:
It initializes some services. Yes we can put all these components to classrooms page too and completely remove the root component also but I don't think we should do that as that will break the existing pattern.

nikitaevg replied:
Consistency is a good argument, let's keep it then

nikitaevg commented on core/templates/pages/classrooms-page/classrooms-page.component.html:
```diff
@@ -0,0 +1,277 @@
+    </div>
+  </div>
+
+  <div class="oppia-classrooms-search-container">
```
Huh, where's this search container? I haven't seen it in the proof of changes

AFZL210 replied:
It's this one. we have oppia-search-bar component which we show in almost all pages but we might show different text above it according to the page.

we need to update oppia-search-bar component too but it's not related to my change I guess.

Desktop
![Screenshot from 2024-06-22 23-38-04](https://github.com/oppia/oppia/assets/79896602/00c11a65-1ef1-43f3-95ab-8ba181d31cff)

Mobile
![Screenshot from 2024-06-22 23-38-21](https://github.com/oppia/oppia/assets/79896602/740078db-56c6-48f3-92ae-09e507f2493e)

vojtechjelinek reviewed (approved):
LGTM!

AFZL210 committed:
minor change

AFZL210 commented:
@nikitaevg PTAL.

nikitaevg commented on core/controllers/access_validators.py:
```diff
@@ -80,6 +80,41 @@ def get(self) -> None:
+            PageNotFoundException. No public classrooms are present.
+        """
+        classrooms = classroom_config_services.get_all_classrooms()
+        has_public_classrooms = [
+            classroom for classroom in classrooms if classroom.is_published
+        ]
```
Oh, that's not "has_public_classrooms". That's "published_classrooms", because that's an array. You do want "has_public_classrooms" here, but it should be defined differently. I suppose any(map(classrooms, lambda c: c.is_published)) should work.

To emphasize, even though this variable behaves like "has_public_classroom", it's an incorrect name, because it's an array of classrooms.

AFZL210 replied:
Oh yes I agree. Thanks for the explanation.
Done.

nikitaevg commented on core/controllers/access_validators.py:
```diff
@@ -80,6 +80,41 @@ def get(self) -> None:
+            classroom for classroom in classrooms if classroom.is_published
+        ]
+
+        if not feature_flag_services.is_feature_flag_enabled(
```
You forgot to move it to the top of the function. It's a small thing, I agree, but it's a bit better, because conceptually you don't need to look at classrooms if the feature flag is disabled. Suppose there's a bug above in the function which causes an exception. If you move the if with the feature flag above, you can "remove" that bug from production just by turning off the feature flag.

AFZL210 replied:
Done.

nikitaevg commented on core/templates/pages/classrooms-page/classrooms-page.component.ts:
```diff
@@ -0,0 +1,90 @@
+})
+export class ClassroomsPageComponent {
+  classroomSummaries: ClassroomSummaryDict[] = [];
+  haveAtleastOnePrivateClassroom: boolean = false;
```
let's rename this accordingly to the backend part

AFZL210 replied:
Done.

nikitaevg reviewed (approved):
Please address remaining comments before merging

AFZL210 committed:
refactor code

AFZL210 committed:
Merge branch 'develop' into fix#19849-classrooms-page

AFZL210 committed:
Merge branch 'develop' into fix#19849-classrooms-page

AFZL210 committed:
Merge branch 'develop' into fix#19849-classrooms-page
